File: CCC/2021/S4.cpp
#include <bits/stdc++.h>

using namespace std;

#define INF 0x3f3f3f3f
typedef pair<int,int> pii;

int main() {
  
  ios_base::sync_with_stdio(false);
  cin.tie(NULL);
  
  int n,w,d; cin >> n >> w >> d;

  vector<vector<int>> adj(n+1,vector<int>());

  for(int i = 0; i < w; i ++){
    int x,y; cin >> x >> y;

    adj[y].push_back(x);
  }

  queue<int> q;
  vector<bool> visited(n+1,false);
  vector<int> distances(n+1,INF);

  q.push(n);
  visited[n] = true;
  distances[n] = 0;

  while(!q.empty()){
    int node = q.front();
    q.pop();

    for(int child:adj[node]){
      if(!visited[child]){
        q.push(child);
        visited[child] = true;
        distances[child] = distances[node] + 1;
      }
    }
  }

  vector<int> subway(n,-1);

  for(int i = 0; i < n; i++){
    cin >> subway[i];
  }

  multiset<pii> times;

  for(int i = 0; i < n; i++){times.insert(make_pair(i+distances[subway[i]],subway[i]));}

  for(int i = 0; i < d; i++){
    int x,y; cin >> x >> y;

    times.erase(make_pair(distances[subway[x-1]]+x-1,subway[x-1]));

    

    times.insert(make_pair(y-1+distances[subway[x-1]],subway[x-1]));

    times.erase(make_pair(distances[subway[y-1]]+y-1,subway[y-1]));
    

    times.insert(make_pair(x-1+distances[subway[y-1]],subway[y-1]));

    pii ans = *times.begin();

    cout << ans.first << "\n";

    int temp = subway[x-1];
    subway[x-1] = subway[y-1];
    subway[y-1] = temp;

    

    
    

  }
  
}

File: CCC/2012/S4.py
from queue import Queue


while True:
  n = int(input())

  if n == 0:
    break

  s = input().split(" ")
  
  if n == 1:
    
    print(0)
    continue

  

  

  game = ""
  for i in s:
    game += i
    game += ","

  game = game[:len(game)-1]

  check = ""

  for i in range(1,n+1):
    check += str(i) + ","

  check = check[:len(check) - 1]

  q =Queue()

  visited = set()

  visited.add(game)

  q.put((game,0))

  b = True
  while not q.empty():
    node,d = q.get()




    l = node.split(",")


    if node == check:
      print(d)
      b = False
      break
        

    

    for i in range(n):
      if i - 1 > -1 and int(l[i][0]) != 0:

        

        if int(l[i-1][0]) > int(l[i][0]) or int(l[i-1][0] == "0"):

          

          child = ",".join(l[:i-1])

          

          if i-1 > 0:
            child += ","

          if l[i-1][0] == "0":
            child += l[i][0] + ","
          else:
            child += l[i][0] + l[i-1] + ","
          
          if len(l[i]) == 1:
            child += "0,"
            
          else:
            child += l[i][1:] + ","

          child += ",".join(l[i+1:])

          if child[-1] == ",":
            child = child[:len(child) - 1]

          if child not in visited:
            
            visited.add(child)
            q.put((child,d+1))
            
      if i + 1 < len(l) and int(l[i][0]) != 0:

        
        if int(l[i+1][0]) > int(l[i][0]) or l[i+1][0] == "0":

          

          

      
          child = ",".join(l[:i])

          if len(l[:i]) > 0:

            child += ","

          

          if len(l[i]) == 1:
            child += "0,"
            
          else:
            child += l[i][1:] + ","

          if l[i+1][0] == "0":
            child += l[i][0] + ","
          else:
            child += l[i][0] + l[i+1] + ","
          
          
          
          

          child += ",".join(l[i+2:])

          if child[-1] == ",":
            child = child[:len(child) - 1]

          if child not in visited:
            visited.add(child)
            q.put((child,d+1))

  if b:
    print("IMPOSSIBLE")
